go-controller/pkg/clustermanager/status_manager/resource_manager.go

```diff
@@ -0,0 +1,68 @@
+package status_manager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	adminpolicybasedrouteapi "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1"
+	adminpolicybasedrouteclientset "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1/apis/clientset/versioned"
+	adminpolicybasedroutelisters "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1/apis/listers/adminpolicybasedroute/v1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ktypes "k8s.io/apimachinery/pkg/types"
+)
+
+type resourceManager[T any] interface {
+	// cluster-scoped resources should ignore namespace
+	get(namespace, name string) (T, error)
+	// cluster-scoped resources should ignore namespace
+	patch(ctx context.Context, namespace, name string, pt ktypes.PatchType, data []byte, opts metav1.PatchOptions,
+		subresources ...string) (T, error)
+	// statusChanged is used to decide if an event should trigger status update
+	statusChanged(oldObj, newObj T) bool
+	// updateStatus should handle nil in case of pointer type, and return (patchBytes, needsUpdate).
+	updateStatus(obj T) ([]byte, bool)
+}
+
+type apbRouteManager struct {
+	lister adminpolicybasedroutelisters.AdminPolicyBasedExternalRouteLister
+	client adminpolicybasedrouteclientset.Interface
+}
+
+func newAPBRouteManager(lister adminpolicybasedroutelisters.AdminPolicyBasedExternalRouteLister, client adminpolicybasedrouteclientset.Interface) *apbRouteManager {
+	return &apbRouteManager{
+		lister: lister,
+		client: client,
+	}
+}
+
+func (m *apbRouteManager) get(namespace, name string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, error) {
+	return m.lister.Get(name)
+}
+
+func (m *apbRouteManager) patch(ctx context.Context, namespace, name string, pt ktypes.PatchType, data []byte, opts metav1.PatchOptions,
+	subresources ...string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, error) {
+	return m.client.K8sV1().AdminPolicyBasedExternalRoutes().Patch(ctx, name, pt, data, opts, subresources...)
+}
+
+func (m *apbRouteManager) statusChanged(oldObj, newObj *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) bool {
+	return !reflect.DeepEqual(oldObj.Status.Messages, newObj.Status.Messages)
+}
+
+func (m *apbRouteManager) updateStatus(route *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) ([]byte, bool) {
+	if route == nil || len(route.Status.Messages) == 0 {
+		return nil, false
+	}
+	newStatus := adminpolicybasedrouteapi.SuccessStatus
+	for _, message := range route.Status.Messages {
```

Comment:
Is this iterating over the latest message from each zone or does it go through all previous messages as well? Maybe I'm confused and what really happens is that each zone is only updating its single entry in the set.

Reply:
that is exactly what happens, every zone owns one message in the list with the latest status